python/pyarrow/dataset.py

```diff
@@ -232,104 +228,260 @@ def _ensure_format(obj):
         raise ValueError("format '{}' is not supported".format(obj))
 
 
-def factory(path_or_paths, filesystem=None, partitioning=None,
-            format=None):
+def _ensure_multiple_sources(paths, filesystem=None):
     """
-    Create a factory which can be used to build a Dataset.
+    Treat a list of paths as files belonging to a single file system
+
+    If the file system is local then also validates that all paths
+    are referencing existing *files* otherwise any non-file paths will be
+    silently skipped (for example on a remote filesystem).
```

Comment:
An option would also be to punt on this one for now (we need to fix this on the C++ anyway, to have an option to not ignore those files when scanning, there is an open JIRA for it). 

I am not sure what the cost is of doing it here upfront in python, but maybe it is fine for `dataset()`, since you can still use the lower level `FileSystemDataset` constructor to avoid it.

Reply:
You mean to check this unconditionally for remote filesystems as well?

Reply:
Because we can pass a flag to validate the file paths set to true by default.

Reply:
I meant to just leave this to the C++ factory behaviour (which is right now to ignore them, but we are going to fix that on that level -> https://issues.apache.org/jira/browse/ARROW-7673, https://issues.apache.org/jira/browse/ARROW-8283

Reply:
Sure, my intention was to provide better error messages at least locally until this feature is not available in the C++ implementation.

Reply:
I don't really have a good intuition on how expensive this is, so I am fine with keeping as is (the behaviour might change again though once we fix this on the c++ side)